tfx/components/example_validator/executor_test.py

```diff
@@ -19,55 +19,153 @@
+    skewed_data = pd.read_csv(original_csv_path)
+    skewed_data = skewed_data.loc[skewed_data['payment_type'] == 'Cash']
+
+    # Write the skewed df to .csv
+    tmp_dir = tempfile.mkdtemp()
+    skewed_dir = os.path.join(tmp_dir, 'skewed')
+    os.mkdir(skewed_dir)
+    skewed_path = os.path.join(skewed_dir, 'skewed_data.csv')
+    skewed_data.to_csv(skewed_path)
+
+    # TFX pipeline components to generate a schema and tfrecords file for the skewed data
+    examples = external_input(skewed_dir)
+    example_gen = CsvExampleGen(input=examples)
+    statistics_gen = components.StatisticsGen(
+        examples=example_gen.outputs['examples'])
+    schema_gen = components.SchemaGen(
+        statistics=statistics_gen.outputs['statistics'],
+        infer_feature_shape=False)
+
+    p = pipeline.Pipeline(
+        pipeline_name='skewed_chicago_taxi_beam',
+        pipeline_root=tmp_dir,
+        components=[example_gen, statistics_gen, schema_gen],
+        metadata_connection_config=metadata.sqlite_metadata_connection_config(
+            os.path.join(tmp_dir, 'metadata.db')))
+
+    BeamDagRunner().run(p)
+
+    # Construct the skewed statistics artifact from pipeline outputs
+    component_dir = os.path.join(tmp_dir, 'StatisticsGen/statistics')
+    component_num = os.listdir(component_dir)[0]
+    skewed_statistics = standard_artifacts.ExampleStatistics()
+    skewed_statistics.uri = os.path.join(component_dir, component_num, 'eval')
+
+    # Set a skew comparator for the 'payment_type' feature
+    tmp_schema_pbtxt_dir = os.path.join(tmp_dir, 'SkewedSchemaGen')
+    os.mkdir(tmp_schema_pbtxt_dir)
+    tmp_schema_pbtxt_path = os.path.join(tmp_schema_pbtxt_dir, 'schema.pbtxt')
+    original_schema_pbtxt_path = os.path.join(self.schema_artifact.uri,
+                                              'schema.pbtxt')
+    shutil.copy2(original_schema_pbtxt_path, tmp_schema_pbtxt_path)
+
+    with open(tmp_schema_pbtxt_path, "r+") as f:
+      contents = f.read()
+      proto_contents = text_format.Parse(contents, schema_pb2.Schema())
+
+      for feature in proto_contents.feature:
+        if feature.name == "payment_type":
+          feature.skew_comparator.infinity_norm.threshold = 0.001
+
+      contents = text_format.MessageToString(proto_contents)
+      f.seek(0)
+      f.write(contents)
+      f.truncate()
+
+    self.schema_artifact.uri = tmp_schema_pbtxt_dir
+
+    input_dict = {
+        executor.STATISTICS_KEY: [self.eval_stats_artifact],
+        executor.SCHEMA_KEY: [self.schema_artifact],
+        executor.TRAINING_STATISTICS_KEY: [skewed_statistics],
+    }
+
+    output_dict = {
+        executor.ANOMALIES_KEY: [self.validation_output],
+    }
+
+    exec_properties = {}
+
+    # Generate anomalies and test that tfdv detected the skew
+    anomalies = self._test_do(input_dict, output_dict, exec_properties)
+    assert 'COMPARATOR_L_INFTY_HIGH' in str(anomalies.anomaly_info)
```

Comment:
Use [`self.assertIn()`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertIn). Note that the keyword assert is different from the assert methods.

Reply:
Done, I'll use the prebuilt assert methods from now on.